[MLIR][LMHLO-Affine] Add PadOp lowering from lmhlo to Affine + Add verifier for lmhlo.Pad op

avarmapml opened the pull request:
This PR adds two commits :-

**Commit 1:**
-- Adds verifier for lmhlo.Pad op which checks :- 
   1. Output and operand rank matches.
   2. Padding configurations are provided for each dimension.
   3. Padding configurations when applied to operand lead to the provided
       output shape.

**Commit 2:**
-- Adds lowering of PadOp from lmhlo -> Affine as a part of `lhlo-legalize-to-affine` pass.
-- The limitation of the implementation is that it expects interior padding to be 0
   and edge padding to be >= 0.

Signed-off-by: Abhishek Varma <abhishek.varma@polymagelabs.com>

avarmapml committed:
[MLIR][LMHLO-Affine] Add PadOp lowering from lmhlo to Affine

-- This commit adds lowering of PadOp from lmhlo -> Affine as a part
   of `lhlo-legalize-to-affine` pass.
-- The limitation of the implementation is that it expects interior padding to be 0
   and edge padding to be >= 0.

Signed-off-by: Abhishek Varma <abhishek.varma@polymagelabs.com>

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/lhlo/transforms/lhlo_legalize_to_affine.cc:
```diff
@@ -462,6 +462,96 @@ class GatherOpConverter : public OpRewritePattern<GatherOp> {
+
+    SmallVector<int64_t> edge_padding_low(
+        op.edge_padding_low().getValues<int64_t>().begin(),
+        op.edge_padding_low().getValues<int64_t>().end());
```
You're querying a map twice with these accessors (`edge_padding_low`, `edge_padding_high`, ...), can you access these ahead of time?

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/lhlo/transforms/lhlo_legalize_to_affine.cc:
```diff
@@ -462,6 +462,96 @@ class GatherOpConverter : public OpRewritePattern<GatherOp> {
+    ShapedType operand_type = operand.getType().dyn_cast<ShapedType>();
+    ShapedType output_type = output.getType().dyn_cast<ShapedType>();
+    unsigned rank = operand_type.getRank();
+    assert(output_type.getRank() == rank);
```
Is the verifier for this op forbidding unranked input?

avarmapml replied:
There is no verifier for `lmhlo.Pad` and I've made changes to assert if the operands `hasRank()`.

joker-eph replied:
If there is no verifier, then this assert could fire, couldn't it?

avarmapml replied:
Yes @joker-eph , it can and instead of asserting on it I've included a check which bails out on lowering instead by returning `failure()`.

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/lhlo/transforms/lhlo_legalize_to_affine.cc:
```diff
@@ -462,6 +462,96 @@ class GatherOpConverter : public OpRewritePattern<GatherOp> {
+        op.interior_padding().getValues<int64_t>().end());
+
+    assert(edge_padding_low.size() == rank && edge_padding_high.size() == rank &&
+           interior_padding.size() == rank);
```
Same question as above: is this covered by the verifier?

avarmapml replied:
Likewise. There is no verifier checks for `lmhlo.Pad` regarding `interior_padding` and `edge_padding_*` attributes.

joker-eph replied:
OK, I guess another question then may be: **should** there be a verifier for these?

avarmapml replied:
Yes, in my opinion, there should be a verifier implemented for the these.

As per : https://www.tensorflow.org/xla/operation_semantics#pad , it is required by padding configurations to specify the amount of edge padding and the interior padding for `each` dimension.

We don't have verifier for this both on `mhlo` and `lhlo`.

I can implement it for both and include it as part of this PR.

Let me know if we should do that OR get rid of these checks from this PR since they `should` be taken care by a verifier and later on raise a PR for verifiers.

joker-eph replied:
If you can introduce the verifier and remove the checks that would be perfect, thanks!

avarmapml replied:
Hi @joker-eph 

I've included a verifier for `lmhlo.Pad` and all this logic will now be taken care by the verifier itself.
I've included another verifier check for the padding configurations too which check for the expected output shape after paddings are applied.

You may re-review this now.

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/lhlo/transforms/lhlo_legalize_to_affine.cc:
```diff
@@ -462,6 +462,96 @@ class GatherOpConverter : public OpRewritePattern<GatherOp> {
+      // Only handle interior padding being zero for now.
+      if (interior_padding[i] != 0)
+        return rewriter.notifyMatchFailure(op,
+                                           "expected zero interior padding");
```
Can you handle the failure cases as early as possible? 

We should do this before starting to populate vectors, etc.

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/lhlo/transforms/lhlo_legalize_to_affine.cc:
```diff
@@ -462,6 +462,107 @@ class GatherOpConverter : public OpRewritePattern<GatherOp> {
+                                          interior_padding_it.end());
+
+    assert(edge_padding_low.size() == rank &&
+           edge_padding_high.size() == rank && interior_padding.size() == rank);
```
same assert problem as above.

avarmapml replied:
Instead of asserting on it I've included a check which bails out on lowering instead by returning `failure()` if the check fails.

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/lhlo/transforms/lhlo_legalize_to_affine.cc:
```diff
@@ -462,6 +462,107 @@ class GatherOpConverter : public OpRewritePattern<GatherOp> {
+    SmallVector<int64_t> edge_padding_high(edge_padding_high_it.begin(),
+                                           edge_padding_high_it.end());
+    SmallVector<int64_t> interior_padding(interior_padding_it.begin(),
+                                          interior_padding_it.end());
```
We should move the the copies **after** the loop that does the early-exit checks.

`for (auto paddings : llvm::zip(edge_padding_low_it, edge_padding_high_it, interior_padding_it))`

(also the suffix `_it` is misleading: these are ranges and not iterators)

avarmapml replied:
This was useful. Thanks.
I've now moved all the early-exit checks prior to assigning/initializing any variables.

avarmapml commented:
Hi @joker-eph 
I've addressed your review comments. You may re-review now.

avarmapml committed:
[MLIR][LHLO] Add verifier for Pad op

-- This commit adds verifier for lhlo.Pad op which checks the following :-
   1. Output and operand rank matches.
   2. Padding configurations are provided for each dimension.
   3. Padding configurations when applied to operand lead to the provided
      output shape.

Signed-off-by: Abhishek Varma <abhishek.varma@polymagelabs.com>

avarmapml commented:
The build failure seems unrelated (Python version mismatch).

CC: @joker-eph

avarmapml commented:
Hi @joker-eph 

The build failures look unrelated to me. Can you please look into it?

joker-eph commented on tensorflow/compiler/mlir/hlo/lib/Dialect/lhlo/IR/lhlo_ops.cc:
```diff
@@ -228,6 +228,67 @@ static LogicalResult Verify(CustomCallOp op) {
+  // dimension is calculated as :-
+  //     low_padding + operand_dim_size + total_interior_padding + high_padding
+  //  where, total_interior_padding = (operand_dim_size - 1) * interior_padding.
+  for (auto paddings : llvm::enumerate(llvm::zip(
```
The linter is complaining and want `const auto &` here
```suggestion
  for (const auto &paddings : llvm::enumerate(llvm::zip(
```

avarmapml replied:
Oh okay. Thank you for helping in the change! :)

There is another at : https://github.com/tensorflow/tensorflow/blob/8e2e96669bdd0b50e03e8552e9e1cfe242c0c229/tensorflow/compiler/mlir/hlo/lib/Dialect/lhlo/transforms/lhlo_legalize_to_affine.cc#L502

Does this need to be changed too? The build failures still exists : `//tensorflow/python/data/experimental/kernel_tests/service:fault_tolerance_test FAILED`

joker-eph committed:
Update tensorflow/compiler/mlir/hlo/lib/Dialect/lhlo/IR/lhlo_ops.cc
